### ISSUE TITLE
improve model strcture

### ISSUE THREAD

**Issue by Shinichi0713** [Shinichi0713/image-ai-models#8]
## puropose
to improve accuracy of difference detection

## content to change
1. feature piramid structure
to get the shallow shape information, abstract the shallow shape feature from first block of model.
use the feature information toward decorder.
2. GAM
make the deformable information.
use GAM mechanizm.

**Comment by Shinichi0713** [comment]
dpr = [x.item() for x in torch.linspace(0, drop_path_rate, sum(depths))]
        cur = 0
        self.block1 = nn.ModuleList([Block(
            dim=embed_dims[0], num_heads=num_heads[0], mlp_ratio=mlp_ratios[0], qkv_bias=qkv_bias, qk_scale=qk_scale,
            drop=drop_rate, attn_drop=attn_drop_rate, drop_path=dpr[cur + i], norm_layer=norm_layer,
            sr_ratio=sr_ratios[0])
            for i in range(depths[0])])

**Comment by Shinichi0713** [comment]
class OverlapPatchEmbed(nn.Module):
    def __init__(self, img_size=224, patch_size=7, stride=4, in_chans=3, embed_dim=768):
        super().__init__()
        # 画像サイズとパッティサイズを2次元に変換
        img_size = to_2tuple(img_size)
        patch_size = to_2tuple(patch_size)

        self.img_size = img_size
        self.patch_size = patch_size
        self.H, self.W = img_size[0] // patch_size[0], img_size[1] // patch_size[1]
        self.num_patches = self.H * self.W      # パッチ数を計算
        self.proj = nn.Conv2d(in_chans, embed_dim, kernel_size=patch_size, stride=stride,
                              padding=(patch_size[0] // 2, patch_size[1] // 2))
        self.norm = nn.LayerNorm(embed_dim)

        self.apply(self._init_weights)

    # ここは原文コードのまま
    def _init_weights(self, m):
        if isinstance(m, nn.Linear):
            trunc_normal_(m.weight, std=.02)
            if isinstance(m, nn.Linear) and m.bias is not None:
                nn.init.constant_(m.bias, 0)
        elif isinstance(m, nn.LayerNorm):
            nn.init.constant_(m.bias, 0)
            nn.init.constant_(m.weight, 1.0)
        elif isinstance(m, nn.Conv2d):
            fan_out = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
            fan_out //= m.groups
            m.weight.data.normal_(0, math.sqrt(2.0 / fan_out))
            if m.bias is not None:
                m.bias.data.zero_()

    # 出力は特徴マップと、分割サイズ
    def forward(self, x):
        # pdb.set_trace()
        x = self.proj(x)
        _, _, H, W = x.shape
        # 2次以降を1次元化して転置
        x = x.flatten(2).transpose(1, 2)
        x = self.norm(x)

        return x, H, W

**Comment by Shinichi0713** [comment]
process 1 is complete

**Comment by Shinichi0713** [comment]
process2 is done.

**Comment by Shinichi0713** [comment]
STM dpm't work well
→remove that from model

then, loss decrease become larger...

**Comment by Shinichi0713** [comment]
![Image](https://github.com/user-attachments/assets/6f9c292a-e123-4ce7-84b5-39c5593fc2dc)

**Comment by Shinichi0713** [comment]
![Image](https://github.com/user-attachments/assets/572eabe6-9726-43cb-a656-a382e299440c)

**Comment by Shinichi0713** [comment]
with seeing feature map, i can find that model can recognize shape feature.
if decorder can get the information clearly, ai can recognize.
main reason not recognize well exists in the truth that decorder can't work well.
so that, i investigate how to get more clear output.

**Comment by Shinichi0713** [comment]
output decorder image

![Image](https://github.com/user-attachments/assets/d34796ae-bdfb-4699-8dc1-0dbe9cffb44b)

in this image, the emphasised shape is just only wall or characters.
actual difference isn't emphasized.

if learning is proceeded, observe this trend.

**Comment by Shinichi0713** [comment]
ai can be becoming to recognize the difference

![Image](https://github.com/user-attachments/assets/b42efeaf-9a92-4813-8d8e-87d5f8ecc8a4)

**Comment by Shinichi0713** [comment]
may be done
